// _@chakra-ui_drawer@1.0.0-next.3@@chakra-ui/drawer/dist/cjs/drawer.js
"use strict";

exports.__esModule = true;
exports.Drawer = Drawer;
exports.DrawerCloseButton = exports.DrawerFooter = exports.DrawerHeader = exports.DrawerBody = exports.DrawerOverlay = exports.DrawerContent = void 0;

var React = _interopRequireWildcard(require("react"));

var _transition = require("@chakra-ui/transition");

var _modal = require("@chakra-ui/modal");

exports.DrawerBody = _modal.ModalBody;
exports.DrawerHeader = _modal.ModalHeader;
exports.DrawerFooter = _modal.ModalFooter;
exports.DrawerCloseButton = _modal.ModalCloseButton;

var _system = require("@chakra-ui/system");

var _utils = require("@chakra-ui/utils");

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

var TransitionContext = /*#__PURE__*/React.createContext({
  content: {},
  overlay: {}
});

if (_utils.__DEV__) {
  TransitionContext.displayName = "TransitionContext";
}

var useTransitionContext = function useTransitionContext() {
  return React.useContext(TransitionContext);
};

function DrawerTransition(props) {
  var inProp = props["in"],
      children = props.children,
      placement = props.placement;
  return /*#__PURE__*/React.createElement(_transition.Slide, {
    "in": inProp,
    placement: placement
  }, function (contentStyle) {
    return /*#__PURE__*/React.createElement(_transition.Fade, {
      "in": inProp
    }, function (overlayStyle) {
      return children({
        content: contentStyle,
        overlay: overlayStyle
      });
    });
  });
}

if (_utils.__DEV__) {
  DrawerTransition.displayName = "DrawerTransition";
}

function Drawer(props) {
  var isOpen = props.isOpen,
      onClose = props.onClose,
      _props$placement = props.placement,
      placement = _props$placement === void 0 ? "right" : _props$placement,
      children = props.children,
      rest = _objectWithoutPropertiesLoose(props, ["isOpen", "onClose", "placement", "children"]);

  return /*#__PURE__*/React.createElement(DrawerTransition, {
    "in": isOpen,
    placement: placement
  }, function (styles) {
    return /*#__PURE__*/React.createElement(TransitionContext.Provider, {
      value: styles
    }, /*#__PURE__*/React.createElement(_modal.Modal, _extends({
      isOpen: true,
      onClose: onClose
    }, rest), children));
  });
}

if (_utils.__DEV__) {
  Drawer.displayName = "Drawer";
}

var DrawerContent = /*#__PURE__*/(0, _system.forwardRef)(function DrawerContent(props, ref) {
  var _useTransitionContext = useTransitionContext(),
      styles = _useTransitionContext.content;

  return /*#__PURE__*/React.createElement(_modal.ModalContent, _extends({
    ref: ref,
    position: "fixed",
    style: styles,
    marginTop: "0",
    marginBottom: "0",
    borderRadius: "0"
  }, props));
});
exports.DrawerContent = DrawerContent;

if (_utils.__DEV__) {
  DrawerContent.displayName = "DrawerContent";
}

var DrawerOverlay = /*#__PURE__*/(0, _system.forwardRef)(function DrawerOverlay(props, ref) {
  var _useTransitionContext2 = useTransitionContext(),
      styles = _useTransitionContext2.overlay;

  return /*#__PURE__*/React.createElement(_modal.ModalOverlay, _extends({
    style: styles,
    transition: "all 0.2s",
    ref: ref
  }, props));
});
exports.DrawerOverlay = DrawerOverlay;

if (_utils.__DEV__) {
  DrawerOverlay.displayName = "DrawerOverlay";
}
